File: fern/authentication/integrations.py
from typing import Any
from composio_openai import ComposioToolSet, App, Action
from composio.client.collections import IntegrationModel
from composio.client.collections import AuthSchemeType

toolset = ComposioToolSet()


def has_integration(app: App) -> bool:
    """
    Check if an integration exists for the given app.
    """
    integrations = toolset.get_integrations(app=app)
    return len(integrations) > 0


def create_integration_if_not_exists(
    app: App,
    auth_mode: AuthSchemeType,
    auth_config: dict,
    use_composio_oauth_app: bool = False,
    force_new_integration: bool = False,
) -> IntegrationModel:
    """
    Create an integration if it doesn't exist.
    """
    if not force_new_integration and has_integration(app):
        integrations = toolset.get_integrations(app=app)
        return integrations[0]
        
    integration = toolset.create_integration(
        app=app,
        auth_mode=auth_mode,
        auth_config=auth_config,
        use_composio_oauth_app=use_composio_oauth_app,
    )
    return integration

print(has_integration(App.GMAIL))

app = toolset.client.apps.get(name="gmail")
expected_params = toolset.fetch_expected_integration_params(
    app=app,
    auth_scheme="OAUTH2",
)
print(expected_params)

integration = create_integration_if_not_exists(
    app=App.GMAIL,
    auth_mode="OAUTH2",
    use_composio_oauth_app=True,
)

File: fern/authentication/integrations.ts
import { OpenAIToolSet } from "composio-core";
const composioToolset = new OpenAIToolSet();

const userId = "00000000-0000-0000-0000-000000000002";
const integrations = await composioToolset.integrations.list({
  appName: "github",
});

interface AuthConfig {
  clientId?: string;
  clientSecret?: string;
  [key: string]: any;
}

async function hasActiveConnection(userId: string, app: string) {
  try {
    const entity = await composioToolset.getEntity(userId);
    const connection = await entity.getConnection({ app });
    if (connection.status === "ACTIVE") {
      console.log(`Connection already exists for user ${userId}`);
      return true;
    }
    return false;
  } catch (error) {
    // Check if the error is about not finding a connection
    if (error instanceof Error && 
        error.message?.includes("Could not find a connection") ||
        (error as any)?.errCode === "SDK::NO_CONNECTED_ACCOUNT_FOUND") {
      return false;
    }
    // Re-throw other errors
    throw error;
  }
}

const isConnected = await hasActiveConnection(userId, "github");
if (isConnected) {
  console.log("Connection already exists for this user");
} else {
  console.log("\nSetting up new connection...");
}

async function requestConnectionParams(app: string, authScheme: string) {
  const connectionParams = await composioToolset.apps.getRequiredParamsForAuthScheme({
    appId: app,
    authScheme: authScheme,
  });
  return connectionParams;
}

const connectionParams = await requestConnectionParams("gmail", "BEARER_TOKEN");
console.log(connectionParams);

process.exit(0);

async function createIntegrationIfNotExists(
  app: string,
  name: string,
  authScheme:
    | "OAUTH2"
    | "OAUTH1"
    | "OAUTH1A"
    | "API_KEY"
    | "BASIC"
    | "BEARER_TOKEN"
    | "GOOGLE_SERVICE_ACCOUNT"
    | "NO_AUTH"
    | "BASIC_WITH_JWT"
    | "COMPOSIO_LINK",
  authConfig: AuthConfig,
  useComposioAuth: boolean = false
) {
  try {
    // Get existing integrations for the app
    const integrations = await composioToolset.integrations.list({
      appName: app,
    });
    // Return first existing integration if found
    if (
      integrations &&
      Array.isArray(integrations) &&
      integrations.length > 0
    ) {
      return integrations[0];
    }

    // Create new integration if none exists
    const integration = await composioToolset.integrations.create({
      name: name,
      appUniqueKey: app,
      authScheme: authScheme,
      authConfig: authConfig,
      useComposioAuth: useComposioAuth,
    });
    return integration;
  } catch (error) {
    console.error("Error creating/getting integration:", error);
    throw error;
  }
}

const integration = await createIntegrationIfNotExists("github", "github-ts", "OAUTH2", {
  client_id: "1234567890",
  client_secret: "1234567890",
});

console.log(integration);

const app = composioToolset.apps.get({ appKey: "github" });

const reqParams = await composioToolset.apps.getRequiredParamsForAuthScheme({
  appId: "firecrawl",
  authScheme: "API_KEY",
});

console.log(reqParams);
